derive/src/trait_bounds.rs

```diff
@@ -22,41 +22,54 @@ use syn::{
     Generics,
     Result,
     Type,
+    TypeGenerics,
+    WhereClause,
 };
 
 /// Adds a `TypeInfo + 'static` bound to all relevant generic types including
 /// associated types (e.g. `T::A: TypeInfo`), correctly dealing with
 /// self-referential types.
-pub fn add(input_ident: &Ident, generics: &mut Generics, data: &syn::Data) -> Result<()> {
-    let ty_params_ids = generics
-        .type_params()
+pub fn add<'a>(
+    input_ident: &'a Ident,
+    generics: &'a Generics,
+    data: &'a syn::Data,
+) -> Result<(TypeGenerics<'a>, WhereClause)> {
+    let (_, ty_generics, where_clause) = generics.split_for_impl();
+    let mut where_clause = where_clause.cloned().unwrap_or_else(|| {
```

Comment:
I don't remember exactly what the problem was but I think I had something similar. I found no way around having to clone things either.

Reply:
They main thing was not using the built in `generics.make_where_clause()` which requires `&mut self`, and just copying what it does without requiring an exclusive reference.